contrib/pkg/adm/managedns/enable.go

```diff
@@ -0,0 +1,211 @@
+// DNS globally for the Hive cluster.
+func NewEnableManageDNSCommand() *cobra.Command {
+	opt := &Options{}
+
+	cmd := &cobra.Command{
+		Use:   `enable domain1.example.com domain2.example.com ...`,
+		Short: "Enable managed DNS globally for the Hive cluster.",
+		Long:  longDesc,
+		Args:  cobra.MinimumNArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			log.SetLevel(log.InfoLevel)
+			if err := opt.Complete(cmd, args); err != nil {
+				return
+			}
+			if err := opt.Validate(cmd); err != nil {
+				return
+			}
+			dynClient, err := contributils.GetClient()
+			if err != nil {
+				log.WithError(err).Fatal("error creating kube clients")
+			}
+
+			err = opt.Run(dynClient, args)
+			if err != nil {
+				log.WithError(err).Error("Error")
+			}
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&opt.Cloud, "cloud", cloudAWS, "Cloud provider: aws(default)|gcp)")
+	flags.StringVar(&opt.CredsFile, "creds-file", "", "Cloud credentials file (defaults vary depending on cloud)")
+	return cmd
+}
+
+// Complete finishes parsing arguments for the command
+func (o *Options) Complete(cmd *cobra.Command, args []string) error {
+	o.homeDir = "."
+
+	if u, err := user.Current(); err == nil {
+		o.homeDir = u.HomeDir
+	}
+	return nil
+}
+
+// Validate ensures that option values make sense
+func (o *Options) Validate(cmd *cobra.Command) error {
+	return nil
+}
+
+// Run executes the command
+func (o *Options) Run(dynClient client.Client, args []string) error {
+	if err := apis.AddToScheme(scheme.Scheme); err != nil {
+		return err
+	}
+	rh, err := o.getResourceHelper()
+	if err != nil {
+		return err
+	}
+
+	// Update the current HiveConfig, which should always exist as the operator will
+	// create a default one once run.
+	hc := &hivev1.HiveConfig{}
+	err = dynClient.Get(context.Background(), types.NamespacedName{Name: "hive"}, hc)
+	if err != nil {
+		log.WithError(err).Fatal("error looking up HiveConfig 'hive'")
+	}
+
+	hc.Spec.ManagedDomains = args
+
+	if o.Cloud == cloudAWS {
+		// Apply a secret for credentials to manage the root domain:
+		credsSecret, err := o.generateAWSCredentialsSecret()
+		if err != nil {
+			log.WithError(err).Fatal("error generating manageDNS credentials secret")
+		}
+		log.Infof("created Route53 credentials secret: %s", credsSecret.Name)
+		credsSecret.Namespace = hiveNamespace
+		rh.ApplyRuntimeObject(credsSecret, scheme.Scheme)
+
+		hc.Spec.ExternalDNS = &hivev1.ExternalDNSConfig{
+			AWS: &hivev1.ExternalDNSAWSConfig{
+				Credentials: corev1.LocalObjectReference{Name: manageDNSCredentialsSecret},
+			},
+		}
+	}
+
+	err = dynClient.Update(context.Background(), hc)
+	if err != nil {
+		log.WithError(err).Fatal("error updating HiveConfig")
+	}
+	log.Info("updated HiveConfig")
+
+	// Adding manageDNS to HiveConfig triggers a new rollout of the hiveadmission pods.
+	// To know when it's safe to proceed, we will list all hiveadmission pods and ensure
```

Comment:
It strikes me that it would have been better to check the rollout status of the deployment rather than checking the pods directly.

Reply:
I tried to get it to happen with the deployment but I couldn't see a way, there were only two conditions to work with and minavailable never flipped true due to how it does the rollout. It's likely possible with replicaset but it didn't look simple. I opted for the pods as it's quite simple and gives us clear info on the state we actually want. It also gets past any race conditions where the changes aren't reflected in the deployment/RS if we immediately check right after editing hiveconfig.

Reply:
I was thinking something along the lines of the following.
1) Get the revision of the deployment.
2) Change the hiveconfig.
3) Wait for the deployment to have both a new revision and a condition of Available=true.

Reply:
What I was seeing however was that Available=true throughout the process, as long as there are min replicas ready, it's happy, and the deployment rolls out a new pod before replacing an old one so it was always marked available.

Reply:
Interesting. Maybe it is the Progressing condition then instead. Nevertheless, the code is adequate the way it is now.

Reply:
I do not know what `kubectl rollout` does under the covers.